csfieldguide/static/interactives/algorithm-timer/js/algorithm-timer.js

```diff
@@ -0,0 +1,177 @@
+    updateData()
+  });
+  $('input[name=result-form]').click(function() {
+    resultForm = $('input[name=result-form]:checked').prop('id');
+    updateData()
+  });
+  $('#n-items').on('input', function() {
+    n = $('#n-items').val();
+    if (inputIsValid(n, speed, processors)) {
+      updateData()
+    }
+  });
+  $('#speed').on('input', function() {
+    speed = $('#speed').val();
+    if (inputIsValid(n, speed, processors)) {
+      updateData()
+    }
+  });
+  $('#processors').on('input', function() {
+    processors = $('#processors').val();
+    if (inputIsValid(n, speed, processors)) {
+      updateData()
+    }
+  });
+  $('input[name=time]').click(function() {
+    timeUnits = $('input[name=time]:checked').prop('id');
+    updateData()
+  });
+
+  /** Get the time it will take for the algorithm to finish executing
+   *  and display it. */
+  function updateData() {
+    var result = calculateTimeTaken(complexity, resultForm, n, speed, processors, timeUnits);
+    $('#output').val(result + ' ' + timeUnits);
+  }
+});
+
+
+/** Calculates the time the algorithm will take to finish
+ *  executing in the selected time units */
+function calculateTimeTaken(complexity, resultForm, n, speed, processors, timeUnits) {
+  // can only have whole integers for number of items
+  n = Mathjs.bignumber(Math.round(n));
+  $('#n-items').val(n);
+  speed = Mathjs.bignumber(speed);
+  processors = Mathjs.bignumber(processors);
+  var steps;
+  
+  if (complexity == 'log') {
+    steps = Mathjs.bignumber(Mathjs.log(n));
+  } else if (complexity == 'linear') {
+    steps = Mathjs.bignumber(n);
+  } else if (complexity == 'nlog') {
+    steps = Mathjs.bignumber(n * Mathjs.log(n));
+  } else if (complexity == 'squared') {
+    steps = Mathjs.bignumber(Math.pow(n, 2));
+  } else if (complexity == 'cubed') {
+    steps = Mathjs.bignumber(Math.pow(n, 3));
+  } else if (complexity == 'fourth-power') {
+    steps = Mathjs.bignumber(Math.pow(n, 4));
+  } else if (complexity == 'sixth-power') {
+    steps = Mathjs.bignumber(Math.pow(n, 6));
+  } else if (complexity == 'exponential') {
+    steps = Mathjs.bignumber(Math.pow(2, n));
+  } else if (complexity == 'factorial') {
+    steps = Mathjs.factorial(n);
+  }
+
+  var denominator = Mathjs.multiply(speed, processors);
+  var timeTaken = Mathjs.divide(steps, denominator);
+
+  var timeScale = Mathjs.bignumber(TIME_SCALERS[timeUnits])
+  timeTaken = Mathjs.multiply(timeTaken, timeScale);
+
+  if (resultForm == 'scientific') {
+    timeTaken = Mathjs.format(Mathjs.bignumber(timeTaken), {notation: 'exponential', precision: 3});
+  } else {
+    timeTaken = Mathjs.format(Mathjs.bignumber(timeTaken), {notation: 'fixed', precision: 2});
+  }
+
+  // If there are no significant decimal places don't show any decimal places
+  if (timeTaken % 1 == 0) {
+    timeTaken = Mathjs.format(Mathjs.bignumber(timeTaken), {notation: 'fixed', precision: 0});
+  }
+
+  if (isNaN(timeTaken)) {
```

Comment:
It'd be better to find out the inputs are invalid right at the beginning, rather than after all the calculations

Reply:
The inputs are checked for validity before this function is called (it's not called if they are invalid). I'm not sure I can find a scenario where this if statement will be true anymore. Should I remove it or keep it there as a just-in-case thing?

Reply:
*Right now* it does get hit, see earlier problem. If the difference is `seconds` vs `NaN seconds` I don't think there's much benefit of keeping it just in case